### ISSUE TITLE
Suggestion: Separate Excalidraw canvas from application UI

### ISSUE THREAD

**Issue by tomhicks** [excalidraw/excalidraw#2609]
This would make embedding much more powerful, and let me incorporate my own UI if I wanted to.

For instance, I want to incorporate exaclidraw in my app, but limited to only certain tools, and would like to use my own UI for tool selection.

If this is something that excalidraw thinks would be a useful addition, I would be willing to work on this.

Possible APIs:

Controlled component.
```
<Excalidraw ui={ExcalidrawDefaultUi} defaultUiProps={{
  tool: "pencil",
  canvasBackgroundColor: "black",
  strokeColor: "white",
}}>
```
... or however you specify your tool config.

Other APIs might include:

render props:
```
<Excalidraw renderUi={(props) => <ExcalidrawDefaultUi {...props} />} />
```

hooks:
```
<ExcalidrawContextProvider>
  <ExcalidrawCanvas />
  <MyExcalidrawUi />
</ExcalidrawContextProvider>

function MyExcalidrawUi() {
  const uiControls = useExcalidrawUi()

  return <div>
    <button style={{fontWeight: uiControls.currentTool === "pencil" ? "bold" : "normal"}} onClick={() => uiControls.setTool("pencil")} >Pencil</button>
    <button style={{fontWeight: uiControls.currentTool === "rect" ? "bold" : "normal"}} onClick={() => uiControls.setTool("rect")} >Rectangle</button>
  </div>
}
```

**Comment by ad1992** [comment]
So you want to use Excalidraw for drawing but add on your tools on top of it which can be done using render props.
The other part is to control what all tools(export, save as, etc)  need to be shown, that can be made config based and default the config will show all the existing tools which are shown right now.
We can definitely iterate over the API once we decide to implement this.
@dwelle what do you think?

**Comment by tomhicks** [comment]
Maybe, I think maybe I have confused things with my explanation, so let me try again.

To clarify my understanding of how Excalidraw currently works:

- The "ExcalidrawCanvas"
  - takes as input the excalidraw document, mouse/touch input, currently selected tool
  - outputs the drawing to canvas/SVG and emits events like "finished drawing", "selected element", etc.
- The "ExcalidrawUI"
  - takes as input the current selected item, current selected tool
  - outputs all the panels for opening documents, choosing tools, updating tool/shape properties, etc.

Rather than being able to control it via config, the entire ExcalidrawUI could be switched out for a completely custom UI that fits with my application.

So as long as I make a UI that conforms to the ExcalidrawUI interface, I could choose to render the various UI panels entirely outside of the canvas rectangle, with my own icons, fonts, etc.

I suggested a few ways that could be achieved, but I'm sure there are other ways also.
